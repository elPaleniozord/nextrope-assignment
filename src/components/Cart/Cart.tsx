import { Accordion, Alert, Button, Card, Col, Container, Nav, Row } from "react-bootstrap"
import { Link } from "react-router-dom"
import { useAppDispatch, useAppSelector } from "../../app/hooks"
import { decrementAmount, incrementAmount, removeProduct, selectCart } from "../../features/cart/cartSlice"
import deleteIcon from '../../deleteIcon.svg';

const Cart = () => {
  const cartState = useAppSelector(selectCart)
  const dispatch = useAppDispatch()

  const itemList = cartState.items.map(item => (
    <Card key={`cart-item-${item.id}`}>
      <Card.Header >
        <Row>
          <Col sm={9}>{item.title}</Col>
          <Col sm={3} style={{display: 'flex', justifyContent: 'flex-end', alignItems: 'center'}}>
            <Button style={{width:'31px', height:'31px', margin: '5px'}} size='sm' onClick={() => dispatch(incrementAmount(item.id))}>+</Button>
            {item.amount}
            <Button style={{width:'31px', height:'31px', margin: '5px'}} size='sm' onClick={() => dispatch(decrementAmount(item.id))}>-</Button>
            <Button style={{width:'31px', height:'31px', margin: '5px'}} size='sm' onClick={() => dispatch(removeProduct(item.id))} variant='danger'><img src={deleteIcon} alt="delete" /></Button>
          </Col>
        </Row>
      </Card.Header>
      
      <Accordion.Collapse eventKey='0'>
        <Card.Body>
        <Card.Subtitle>{item.author}</Card.Subtitle>
        <Card.Text>Liczba stron: {item.pages}</Card.Text>
        <Card.Text>Cena: {item.price.toString().replace(/\d\d$/,".$&")}zł</Card.Text>
        </Card.Body>
      </Accordion.Collapse>
    </Card>
  ))
  return (
    <Container>
      <h2>Koszyk</h2>
      {itemList.length > 0 ? (<Accordion>{itemList}</Accordion>) : <Alert>Twój koszyk jest pusty</Alert>}

      <div className="d-flex justify-content-between align-items-center">
        <h3>Wartość zamówienia: {cartState.total.toString().replace(/\d\d$/,".$&")}zł</h3>
        <Nav.Link as={Link} to='/checkout' disabled={itemList.length < 1}><Button disabled={itemList.length < 1}>DALEJ</Button></Nav.Link>
      </div>
      
    </Container>
  )
}

export default Cart