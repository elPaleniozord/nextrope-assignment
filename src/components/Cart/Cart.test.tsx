import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import Cart from './Cart';
import { configureStore } from '@reduxjs/toolkit';
import cartReducer, { addProduct } from '../../features/cart/cartSlice';
import { MemoryRouter } from 'react-router';

const book = {
  "id": 457,
  "title": "Matematyka 1. Podręcznik. Zakres podstawowy",
  "author": "M. Karpiński, M. Dobrowolska, M. Braun, J. Lech",
  "cover_url": "http://localhost:3001/static/cover/book/457.jpg",
  "pages": 280,
  "price": 3200,
  "currency": "PLN"
}

export function renderWithStore (
  ui, {
    preloadedState, 
    store = configureStore({reducer: {cart: cartReducer}, preloadedState}),
    ...renderOptions
  } = {}
) {
  function Wrapper({children}) {
    store.dispatch(addProduct(book))
    return <Provider store={store}>{children}</Provider>
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions})
}

test('should render cart component', () => {
  const { getByText, getAllByText, asFragment } = renderWithStore(
    <MemoryRouter>
      <Cart />
    </MemoryRouter>
  );
  
  expect(asFragment()).toMatchSnapshot()
});
