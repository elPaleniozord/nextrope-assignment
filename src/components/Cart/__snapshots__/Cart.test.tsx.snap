// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render cart component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h2>
      Koszyk
    </h2>
    <div
      class="accordion"
    >
      <div
        class="card"
      >
        <div
          class="card-header"
        >
          <div
            class="row"
          >
            <div
              class="col-sm-9"
            >
              Matematyka 1. Podręcznik. Zakres podstawowy
            </div>
            <div
              class="col-sm-3"
              style="display: flex; justify-content: flex-end; align-items: center;"
            >
              <button
                class="btn btn-primary btn-sm"
                style="width: 31px; height: 31px; margin: 5px;"
                type="button"
              >
                +
              </button>
              1
              <button
                class="btn btn-primary btn-sm"
                style="width: 31px; height: 31px; margin: 5px;"
                type="button"
              >
                -
              </button>
              <button
                class="btn btn-danger btn-sm"
                style="width: 31px; height: 31px; margin: 5px;"
                type="button"
              >
                <img
                  alt="delete"
                  src="deleteIcon.svg"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="accordion-collapse collapse"
        >
          <div
            class="card-body"
          >
            <div
              class="card-subtitle h6"
            >
              M. Karpiński, M. Dobrowolska, M. Braun, J. Lech
            </div>
            <p
              class="card-text"
            >
              Liczba stron: 280
            </p>
            <p
              class="card-text"
            >
              Cena: 32.00zł
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="d-flex justify-content-between align-items-center"
    >
      <h3>
        Wartość zamówienia: 32.00zł
      </h3>
      <a
        class="nav-link"
        href="/checkout"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          DALEJ
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
